plugins/widget/plugin.js

```diff
@@ -3071,35 +3224,85 @@
 	// LEFT, RIGHT, UP, DOWN, DEL, BACKSPACE - unblock default fake sel handlers.
 	var keystrokesNotBlockedByWidget = { 37: 1, 38: 1, 39: 1, 40: 1, 8: 1, 46: 1 };
 
-	// Applies or removes style's classes from widget.
+	// Adds or removes class to `widget.styleDefinition`.
+	// @param {CKEDITOR.plugins.widget} widget instance.
+	// @param {String} CSS class.
+	// @param {Boolean} whenever to remove or add class.
+	function addRemoveClassToStyleDef( widget, className, remove ) {
+		var classes;
+		if ( widget.styleDefinition && widget.styleDefinition.attributes[ 'class' ] ) {
+			classes = widget.styleDefinition.attributes[ 'class' ].split( /\s+/ );
+		} else {
+			classes = [];
+		}
+
+		if ( remove ) {
+			classes = CKEDITOR.tools.array.filter( classes, function( item ) {
+				return item !== className;
+			} );
+		} else {
+			if ( CKEDITOR.tools.indexOf( classes, className ) == -1 ) {
+				classes.push( className );
+			}
+		}
+		widget.styleDefinition = widget.styleDefinition || { attributes: { 'class': '' } };
+		widget.styleDefinition.attributes[ 'class' ] = classes.join( ' ' );
+	}
+
+	// Applies or removes style's attributes from widget.
 	// @param {CKEDITOR.style} style Custom widget style.
-	// @param {Boolean} apply Whether to apply or remove style.
+	// @param {Boolean} apply Whenever to apply or remove style.
 	function applyRemoveStyle( widget, style, apply ) {
 		var changed = 0,
 			classes = getStyleClasses( style ),
 			updatedClasses = widget.data.classes || {},
-			cl;
-
+			cl,
+			styles = style.getDefinition().styles,
+			attributes = style.getDefinition().attributes;
 		// Ee... Something is wrong with this style.
-		if ( !classes )
+		if ( !classes && !styles && !attributes ) {
 			return;
+		}
 
-		// Clone, because we need to break reference.
-		updatedClasses = CKEDITOR.tools.clone( updatedClasses );
+		if ( classes ) {
+			// Clone, because we need to break reference.
+			updatedClasses = CKEDITOR.tools.clone( updatedClasses );
 
-		while ( ( cl = classes.pop() ) ) {
-			if ( apply ) {
-				if ( !updatedClasses[ cl ] )
-					changed = updatedClasses[ cl ] = 1;
-			} else {
-				if ( updatedClasses[ cl ] ) {
+			while ( ( cl = classes.pop() ) ) {
+				if ( apply ) {
+					if ( !updatedClasses[ cl ] )
+						changed = updatedClasses[ cl ] = 1;
+				} else if ( updatedClasses[ cl ] ) {
 					delete updatedClasses[ cl ];
 					changed = 1;
 				}
 			}
+			if ( changed ) {
+				widget.setData( 'classes', updatedClasses );
```

Comment:
Getting rid of data.classes wouldn't break users custom plugins that extend widget?  I'm not sure how to approach this.

Reply:
I'm quite sure that this data is used only internally by our plugin to know, which styles are applied to the widget. I don't even see any info about `data.classes` in our docs. @mlewand, am I right or I'm missing something?

Reply:
https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_plugins_widget.html#method-setData

I've found only this references to `data.classes`.

Reply:
We can't get rid of `data.classes` it has been already exposed. Among all it's used by the `image2` plugin.

Community had already a good time on picking it up, and relying on that. Because of that we should keep it for compatibility reasons.